// @babel/traverse/lib/cache.js
/**
 * Code developed by Isaac Muliro - UI/UX Designer & Developer
 *
 * Usage Guidelines:
 * - Maintain modular structure when adding new features
 * - Use ES6+ syntax standards and some times I built my own modules from sratch
 * - Document any new functions with JSDoc comments
 * - For questions or contributions, contact isaac.muliro@purchase.edu
 * - Last updated: 2025-05-06
 */



"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.clear = clear;
exports.clearPath = clearPath;
exports.clearScope = clearScope;
exports.getCachedPaths = getCachedPaths;
exports.getOrCreateCachedPaths = getOrCreateCachedPaths;
exports.scope = exports.path = void 0;
let pathsCache = exports.path = new WeakMap();
let scope = exports.scope = new WeakMap(); function clear() {clearPath();
  clearScope();
} function clearPath() {exports.path = pathsCache = new WeakMap();
} function clearScope() {exports.scope = scope = new WeakMap();
} function getCachedPaths(path) {const { parent, parentPath
  } = path;
  return pathsCache.get(parent);
} function getOrCreateCachedPaths(node, parentPath) {;let paths = pathsCache.get(node);if (!paths) pathsCache.set(node, paths = new Map());return paths;
}
